core/src/main/java/org/apache/brooklyn/enricher/stock/reducer/Reducer.java

```diff
@@ -0,0 +1,187 @@
+        "A map of parameters to pass into the reducer function");
+   
+    protected Entity producer;
+    protected List<AttributeSensor<S>> subscribedSensors;
+    protected Sensor<T> targetSensor;
+    protected Function<List<S>, T> reducerFunction;
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        Preconditions.checkNotNull(getConfig(SOURCE_SENSORS), "source sensors");
+
+        this.producer = getConfig(PRODUCER) == null ? entity : getConfig(PRODUCER);
+        List<AttributeSensor<S>> sensorListTemp = Lists.newArrayList();
+
+        for (Object sensorO : getConfig(SOURCE_SENSORS)) {
+            AttributeSensor<S> sensor = Tasks.resolving(sensorO).as(AttributeSensor.class).timeout(ValueResolver.REAL_QUICK_WAIT).context(producer).get();
+            Optional<? extends Sensor<?>> foundSensor = Iterables.tryFind(sensorListTemp, 
+                    SensorPredicates.sensorNameEqualTo(sensor.getName()));
+            
+            if(!foundSensor.isPresent()) {
+                sensorListTemp.add(sensor);
+            }
+        }
+        
+        String reducerName = config().get(REDUCER_FUNCTION_TRANSFORMATION);
+        Function<List<S>, T> reducerFunction = (Function) config().get(REDUCER_FUNCTION);
+        if(reducerFunction == null){
+            Map<String, ?> parameters = config().get(PARAMETERS);
+            reducerFunction = createReducerFunction(reducerName, parameters);
+        }
+
+        this.reducerFunction = reducerFunction;
+        Preconditions.checkState(sensorListTemp.size() > 0, "Nothing to reduce");
+
+        for (Sensor<S> sensor : sensorListTemp) {
+            subscribe(producer, sensor, this);
+        }
+
+        subscribedSensors = ImmutableList.copyOf(sensorListTemp);
+    }
+
+    protected abstract Function<List<S>, T> createReducerFunction(String reducerName, Map<String, ?> parameters);
+    
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onEvent(SensorEvent<Object> event) {
+        Sensor<T> destinationSensor = (Sensor<T>) getConfig(TARGET_SENSOR);
+
+        List<S> values = Lists.newArrayList();
+
+        for (AttributeSensor<S> sourceSensor : subscribedSensors) {
+            S resolvedSensorValue = entity.sensors().get(sourceSensor);
+            if (resolvedSensorValue == null) {
+                // only apply function if all values are resolved
+                return;
+            }
+
+            values.add(resolvedSensorValue);
+        }
+
+        Object result = reducerFunction.apply(values);
+
+        if (LOG.isTraceEnabled()) LOG.trace("enricher {} got {}, propagating via {} as {}",
+                new Object[] {this, event, entity, reducerFunction, destinationSensor});
+
+        emit((Sensor<T>)destinationSensor, result);
+    }
+   
+    public static class JoinerReducerFunction<A> implements Function<List<A>, String> {
```

Comment:
Why not use `StringFunctions.joiner(separator)` (which delegates to `Strings.join` for its `apply` method)?